feat: add emoji support

rinzool opened the pull request:
## Why
I really enjoy `md-to-pdf`, and I like using emojis in it, so I propose a feature to convert emojis like `:unicorn:` to an svg image

## How
Based on [this issue](https://github.com/markedjs/marked/issues/233) and [twemoji](https://github.com/twitter/twemoji), I added a custom `marked` extension to convert emojis.
To avoid breaking changes, it is not enabled by default. One should add option `--emoji` to add emoji parsing.

## Disclaimer
I am not familiar with typescript, so I think my code can be improved 
Furthermore I did not currently add test, but I would like to 

Please feel free to discuss about this feature!

simonhaenisch commented:
Hey @rinzool thanks a lot for the PR! I think it would be nice if you could publish this as a `marked-plugin-emoji` (or similar name). Because basically how I want extensibility to look from a user's perspective is this:

```ts
// md-to-pdf.config.js

import { emojiPlugin } from 'marked-plugin-emoji'

export default {
  markedExtensions: [emojiPlugin]
}
```

In this case I'd also be willing to include this as a default extension/plugin (can make a new major version instead of introducing a flag for the feature). I just don't think the code for it should live with `md-to-pdf`, but separate. I just looked at [some `markdown-it` plugins](https://www.npmjs.com/search?q=keywords:markdown-it-plugin) and there's one for emoji there, so this would basically be the same but for `marked`.

If you're not keen to publish/maintain this as a library, please let me know and I can do it.

rinzool commented:
> Hey @rinzool thanks a lot for the PR! I think it would be nice if you could publish this as a `marked-plugin-emoji` (or similar name). Because basically how I want extensibility to look from a user's perspective is this:
> 
> ```ts
> // md-to-pdf.config.js
> 
> import { emojiPlugin } from 'marked-plugin-emoji'
> 
> export default {
>   markedExtensions: [emojiPlugin]
> }
> ```
> 
> In this case I'd also be willing to include this as a default extension/plugin (can make a new major version instead of introducing a flag for the feature). I just don't think the code for it should live with `md-to-pdf`, but separate. I just looked at [some `markdown-it` plugins](https://www.npmjs.com/search?q=keywords:markdown-it-plugin) and there's one for emoji there, so this would basically be the same but for `marked`.
> 
> If you're not keen to publish/maintain this as a library, please let me know and I can do it.

Thank you @simonhaenisch for your message! 
I think this is a great idea to create a seperate plugin for it yes! :+1: 
I will do that, and I will ask for your help and reviews if you are agree :wink: 
I keep you in touch :ok_hand:

simonhaenisch commented:
Ok, btw yesterday I found out that there are some plugins available already, you can have a look here: https://marked.js.org/using_advanced#extensions. The recommended naming is `marked-<plugin-name>`, i.e. `marked-emoji` in your case. You can probably take some inspiration from those plugins, but basically i think you just need an `index.js` that exports an object that follows the structure explained in https://marked.js.org/using_pro#extensions (I guess it's basically what you defined as `emojiExtension` in this PR). Regarding the type, you can use `marked.TokenizerExtension` (you'll have to add `marked` as a dev and peer dependency).

And yeah feel free to ask me for reviews or help with the publishing.

simonhaenisch commented:
@rinzool have you gotten around to making a marked plugin for this already?

rinzool commented:
> @rinzool have you gotten around to making a marked plugin for this already?

@simonhaenisch  I haven't yet began, I  do not have a lot of time currently. Maybe in the next few feeks :crossed_fingers:

rinzool commented:
Hi @simonhaenisch, 
I finally created a marked plugin: https://github.com/rinzool/marked-emoji/tree/feat/init-plugin
For now it is in `1.0.0-SNAPSHOT` version
I'll be glad if you can read the following PR: https://github.com/rinzool/marked-emoji/pull/1

I also updated the current PR to add `marked-emoji` as a default extension for `md-to-pdf`
I added a new `marked_tokenizer_extensions` list in the config because custom extensions cannot be added using `marked.used(markedEmoji)`

I also have a strange behavious to import `markedEmoji`, on `md-to-pdf` I had to to `markedEmoji.markedEmoji` 

It is still under development, please feel free to help me improve the plugin so we can deploy it

simonhaenisch commented:
> I had to to `markedEmoji.markedEmoji`

you have to change to a named import:

```diff
- const markedEmoji = require('marked-emoji')
+ const { markedEmoji } = require('marked-emoji')
```

simonhaenisch commented:
I'll try to have a look at your library some time next week.

simonhaenisch commented on src/lib/get-marked-with-highlighter.ts:
```diff
@@ -12,5 +16,6 @@ export const getMarked = (options: marked.MarkedOptions, extensions: marked.Mark
 		...options,
 	});
 	marked.use(...extensions);
+	marked.use({ extensions: tokenizerExtensions });
```
I think the current `marked_extensions` might be wrong, should probably fix that, instead of adding an extra option?

rinzool replied:
Based on [documentation](https://marked.js.org/using_pro#extensions), it seems that extensions should be passed like this `{extensions: [..]}`
I updated typing and it seems to work like that

simonhaenisch commented on src/lib/config.ts:
```diff
@@ -172,7 +173,8 @@ interface BasicConfig {
 	 */
-	marked_extensions: marked.MarkedExtension[];
+	// marked_extensions: (marked.TokenizerExtension | marked.RendererExtension | (marked.TokenizerExtension & marked.RendererExtension))[];
+	marked_extensions: (marked.TokenizerExtension | marked.RendererExtension | (marked.TokenizerExtension & marked.RendererExtension))[];
```
Isn't this the same as

```suggestion
	marked_extensions: marked.MarkedExtension[];
```

?

rinzool replied:
Apparently no, with `marked.MarkedExtension` I have the following error:
```console
src/lib/get-marked-with-highlighter.ts:14:15 - error TS2322: Type 'MarkedExtension[]' is not assignable to type '(TokenizerExtension | RendererExtension | (TokenizerExtension & RendererExtension))[]'.
  Type 'MarkedExtension' is not assignable to type 'TokenizerExtension | RendererExtension | (TokenizerExtension & RendererExtension)'.
    Type 'MarkedExtension' is not assignable to type 'TokenizerExtension & RendererExtension'.
      Type 'MarkedExtension' is missing the following properties from type 'TokenizerExtension': name, level

14  marked.use({ extensions });

```

simonhaenisch replied:
Ah ok, I just had a look at the types, so there's two options... either we change it to

```suggestion
	marked_extensions: marked.MarkedExtension['extensions'][];
```

or we leave it and change the default config to

```ts
{
  // ...
  marked_extensions: [{ extensions: [markedEmoji] }],
}
```

rinzool replied:
I like the second idea, I just commited it :+1:

simonhaenisch commented on src/lib/config.ts:
```diff
@@ -172,7 +173,7 @@ interface BasicConfig {
 	 * @see https://marked.js.org/using_pro#extensions
 	 */
-	marked_extensions: marked.MarkedExtension[];
+	marked_extensions: marked.MarkedExtension;
```
Hm I'm not really happy with this not being an array anymore however I also find the Marked extension API a bit confusing 😕 

Looking at https://marked.js.org/using_pro#use and the source of `@types/marked` again, and the way it's exposed by Marked, the most correct would be

```suggestion
	marked_extensions: marked.MarkedExtension[];
```

However I've looked at a few other published extension and they usually don't export a `MarkedExtension` but a `TokenizerExtension` or `RendererExtension`, and that's also how it's stated in the [custom extensions](https://marked.js.org/using_pro#extensions) docs.

Thus I think it's better to go with the first idea I had:

```suggestion
	marked_extensions: marked.MarkedExtension['extensions'];
```

and later do

```ts
marked.use({ extensions })
```

I'll open a bug ticket about this in Marked cause there's a discrepancy about this in their docs.

rinzool replied:
I think it is good now :+1: Thank for your help

simonhaenisch commented on src/lib/get-marked-with-highlighter.ts:
```diff
@@ -1,7 +1,7 @@
 import { marked } from 'marked';
 
-export const getMarked = (options: marked.MarkedOptions, extensions: marked.MarkedExtension[]) => {
+export const getMarked = (options: marked.MarkedOptions, extensions: marked.MarkedExtension) => {
```
```suggestion
export const getMarked = (options: marked.MarkedOptions, extensions: marked.MarkedExtension['extensions']) => {
```

simonhaenisch commented on src/lib/get-marked-with-highlighter.ts:
```diff
@@ -11,6 +11,6 @@ export const getMarked = (options: marked.MarkedOptions, extensions: marked.Mark
 		...options,
 	});
-	marked.use(...extensions);
+	marked.use(extensions);
```
```suggestion
	marked.use({ extensions });
```

rinzool commented on src/lib/config.ts:
```diff
@@ -3,6 +3,7 @@ import { GrayMatterOption } from 'gray-matter';
 import { marked } from 'marked';
 import { resolve } from 'path';
 import { FrameAddScriptTagOptions, launch, PDFOptions } from 'puppeteer';
+const { markedEmoji } = require('marked-twemoji');
```
I added a `.d.ts` on `marked-twemoji` :+1:

simonhaenisch reviewed (commented):
LGTM now, but I realized that this would be a breaking change because all unicode emojis are converted with twemoji (initially I thought this would only convert `:<emoji>:` syntax to the actual unicode emoji).

Thus I would either hide this behind a `twemoji` flag, or instead add the instructions to the readme for adding the plugin via config file.

rinzool commented:
@simonhaenisch I can do both! I will tell you when it's done

rinzool commented:
Hi @simonhaenisch, I deployed `marked-twemoji@1.0.0` 
Thanks again for your help :+1: 
You can merge this PR too if you want

rinzool commented:
Hi @simonhaenisch! The pull request looks good to you? If so, you can merge it i think :smiley:

simonhaenisch commented:
Hey @rinzool sorry I haven't gotten around to this yet (I'm on vacation at the moment). Will get this merged in the next couple weeks though.

rinzool commented:
Hi @simonhaenisch 
Do you plan to merge this PR soon? :grin:

simonhaenisch commented:
Sorry completely forgot, I'll try to get it done.

simonhaenisch commented:
Sorry, I've had zero time for this project lately, but it should get better next month.

rinzool commented:
Hi @simonhaenisch, it's been a long time!
I see that you opened an issue about this PR: https://github.com/simonhaenisch/md-to-pdf/issues/228
What do you want to do next? 

I still use my fork until it's merged on your main branch, but I'll be happy if we could update this PR to merge it :wink:
